h/static/scripts/directive/annotation.js

```diff
@@ -0,0 +1,730 @@
+* highlights are valid.
+*
+* @param {object} annotation The annotation to be validated
+*
+*/
+function validate(annotation) {
+  var permissions;
+  var readPermissions;
+  var worldReadable = false;
+  var targets;
+
+  if (!angular.isObject(annotation)) {
+    return;
+  }
+
+  permissions = annotation.permissions || {};
+  readPermissions = permissions.read || [];
+  targets = annotation.target || [];
+
+  if (annotation.tags && annotation.tags.length) {
+    return annotation.tags.length;
+  }
+
+  if (annotation.text && annotation.text.length) {
+    return annotation.text.length;
+  }
+
+  if (readPermissions.indexOf('group:__world__') !== -1) {
+    worldReadable = true;
+  }
+
+  return (targets.length && !worldReadable);
+}
+
+/** Return a human-readable error message for the given server error.
+ *
+ * @param {object} reason The error object from the server. Should have
+ * `status` and, if `status` is not `0`, `statusText` and (optionally)
+ * `data.reason` properties.
+ *
+ * @returns {string}
+ */
+function errorMessage(reason) {
+  var message;
+  if (reason.status === 0) {
+    message = 'Service unreachable.';
+  } else {
+    message = reason.status + ' ' + reason.statusText;
+    if (reason.data.reason) {
+      message = message + ': ' + reason.data.reason;
+    }
+  }
+  return message;
+}
+
+/**
+  * @ngdoc type
+  * @name annotation.AnnotationController
+  *
+  * @property {Object} annotation The annotation view model.
+  * @property {Object} document The document metadata view model.
+  * @property {string} action One of 'view', 'edit', 'create' or 'delete'.
+  * @property {string} preview If previewing an edit then 'yes', else 'no'.
+  * @property {boolean} editing True if editing components are shown.
+  * @property {boolean} isSidebar True if we are in the sidebar (not on the
+  *                               stream page or an individual annotation page)
+  *
+  * @description
+  *
+  * `AnnotationController` provides an API for the annotation directive. It
+  * manages the interaction between the domain and view models and uses the
+  * {@link annotationMapper AnnotationMapper service} for persistence.
+  */
+// @ngInject
+function AnnotationController(
+  $document, $q, $rootScope, $scope, $timeout, $window, annotationUI,
+  annotationMapper, drafts, flash, groups, permissions, session, tags, time) {
+
+  var vm = this;
+
+  var highlight;
+  var isNewAnnotation;
+
+  /** The domain model, contains the currently saved version of the annotation
+   * from the server. */
+  var model;
+
+  var updateDomainModel;
+  var updateDraft;
+  var updateTimestamp;
+
+  /** The view model, contains user changes to the annotation that haven't
+   * been saved to the server yet. */
+  vm.annotation = {};
+
+  vm.action = 'view';
+  vm.document = null;
+  vm.editing = false;
+  vm.isSidebar = false;
+  vm.preview = 'no';
+  vm.timestamp = null;
+
+  model = $scope.annotationGet();
+  if (model.user === undefined) {
```

Comment:
Done. Incidentally this is how I write in Python as well - making use of truthiness and falsiness in boolean expressions, but @nickstenning often requests me to change it to make it explicit (`if foo is False` etc)